services/motion/localizer.go

```diff
@@ -92,3 +95,54 @@ func (m *movementSensorLocalizer) CurrentPosition(ctx context.Context) (*referen
 	pose := spatialmath.NewPose(spatialmath.GeoPointToPoint(gp, m.origin), o)
 	return referenceframe.NewPoseInFrame(m.Name().Name, spatialmath.Compose(pose, m.calibration)), nil
 }
+
+// TwoDLocalizer will check the orientation of the pose of a localizer, and ensure that it is normal to the XY plane.
+// If it is not, it will be altered such that it is (accounting for e.g. an ourdoor base with one wheel on a rock). If the orientation is
+// such that the base is pointed directly up or down (or is upside-down), an error is returned.
+// The alteration to ensure normality to the plane is done by transforming the (0,1,0) vector by the provided orientation, and then
+// using atan2 on the new x and y values to determine the vector of travel that would be followed.
+func TwoDLocalizer(l Localizer) Localizer {
+	return &yForwards2dLocalizer{l}
+}
+
+type yForwards2dLocalizer struct {
+	Localizer
+}
+
+func (y *yForwards2dLocalizer) CurrentPosition(ctx context.Context) (*referenceframe.PoseInFrame, error) {
+	currPos, err := y.Localizer.CurrentPosition(ctx)
+	if err != nil {
+		return nil, err
+	}
+	newPose, err := projectOrientationTo2dRotation(currPos.Pose())
+	if err != nil {
+		return nil, err
+	}
+	newPiF := referenceframe.NewPoseInFrame(currPos.Parent(), newPose)
+	newPiF.SetName(currPos.Name())
+	return newPiF, nil
+}
+
+func projectOrientationTo2dRotation(pose spatialmath.Pose) (spatialmath.Pose, error) {
+	orient := pose.Orientation().OrientationVectorRadians() // OV is easy to check if we are in plane
+	if orient.OZ < 0 {
+		return nil, errors.New("base appears to be upside down, check your movement sensor")
+	}
+	if orient.OX == 0 && orient.OY == 0 {
+		return pose, nil
+	}
+
+	// This is the vector in which the base would move if it had no changed orientation
+	adjPt := spatialmath.NewPoseFromPoint(r3.Vector{0, 1, 0})
+	// This is the vector the base would follow based on the reported orientation, were it unencumbered by things like gravity or the ground
+	newAdjPt := spatialmath.Compose(spatialmath.NewPoseFromOrientation(orient), adjPt).Point()
+	if 1-math.Abs(newAdjPt.Z) < poleEpsilon {
+		if newAdjPt.Z > 0 {
+			return nil, errors.New("base appears to be pointing straight up, check your movement sensor")
+		}
+		return nil, errors.New("base appears to be pointing straight down, check your movement sensor")
+	}
+	// This is the vector across the ground of the above hypothetical vector, projected onto the X-Y plane.
+	theta := -math.Atan2(newAdjPt.Y, -newAdjPt.X)
```

Comment:
this is the same as `math.Atan2(Y, X)` right?

Reply:
It is not
https://go.dev/play/p/UG-c5xLiBn4